gobblin-service/src/main/java/org/apache/gobblin/service/monitoring/SpecStoreChangeMonitor.java

```diff
@@ -0,0 +1,158 @@
+  private Meter deletedSpecs;
+  private Meter unexpectedErrors;
+
+  protected CacheLoader<String, String> cacheLoader = new CacheLoader<String, String>() {
+    @Override
+    public String load(String key) throws Exception {
+      return key;
+    }
+  };
+
+  protected LoadingCache<String, String>
+      specChangesSeenCache = CacheBuilder.newBuilder().expireAfterWrite(10, TimeUnit.MINUTES).build(cacheLoader);
+
+  @Inject
+  protected FlowCatalog flowCatalog;
+
+  @Inject
+  protected GobblinServiceJobScheduler scheduler;
+
+  public SpecStoreChangeMonitor(String topic, Config config, int numThreads) {
+    super(topic, config, numThreads);
+  }
+
+  @Override
+  /*
+  Note that although this class is multi-threaded and will call this message for multiple threads (each having a queue
+  associated with it), a given message itself will be partitioned and assigned to only one queue.
+   */
+  protected void processMessage(DecodeableKafkaRecord message) {
+    String specUri = (String) message.getKey();
+    SpecStoreChangeEvent value = (SpecStoreChangeEvent) message.getValue();
+
+    Long timestamp = value.getTimestamp();
+    String operation = value.getOperationType().name();
+    log.info("Processing message with specUri is {} timestamp is {} operation is {}", specUri, timestamp, operation);
+
+    // If we've already processed a message with this timestamp and spec uri before then skip duplicate message
+    String changeIdentifier = timestamp.toString() + specUri;
+    if (specChangesSeenCache.getIfPresent(changeIdentifier) != null) {
+      return;
+    }
+
+    // If event is a heartbeat type then log it and skip processing
+    if (operation == "HEARTBEAT") {
+      log.debug("Received heartbeat message from time {}", timestamp);
+      return;
+    }
+
+    Spec spec;
+    URI specAsUri = null;
+
+    try {
+      specAsUri = new URI(specUri);
+    } catch (URISyntaxException e) {
+      log.warn("Could not create URI object for specUri {} due to error {}", specUri, e.getMessage());
+    }
+
+    spec = (operation != "DELETE") ? this.flowCatalog.getSpecWrapper(specAsUri) : null;
+
+    // The monitor should continue to process messages regardless of failures with individual messages, instead we use
+    // metrics to keep track of failure to process certain SpecStoreChange events
+    try {
+      // Call respective action for the type of change received
+      AddSpecResponse response;
+      if (operation == "INSERT" || operation == "UPDATE") {
+        response = scheduler.onAddSpec(spec);
+
+        // Null response means the dag failed to compile
+        if (response != null && response.getValue() != null) {
+          log.info("Successfully added spec {} response {}", spec, StringEscapeUtils.escapeJson(response.getValue().toString()));
+          this.successfullyAddedSpecs.mark();
+        } else {
+          log.warn("Failed to add spec {} due to response {}", spec, response);
+          this.failedAddedSpecs.mark();
```

Comment:
what's the difference between failed added specs and unexpected error?

Reply:
I wanted to differentiate between compilation error (an error we do expect to run into) and other runtime errors we don't (for example receiving an operation type that is not expected). The warning message could be sufficient if we don't care distinguishing the count of each but it can provide us better granularity. I don't expect many of the latter case so we can also group count while having the two separate log messages for each case. What do you think?